Adds OpenTelemetry output description (#1625)
grafana/k6-docs: docs/sources/next/results-output/real-time/opentelemetry.md

---
title: 'OpenTelemetry'
description: 'Use the OpenTelemetry output to send test results in OTEL metrics format.'
weight: 00
---

# OpenTelemetry

> k6 can push test run metrics in a [OpenTelemetry metrics format](https://opentelemetry.io/docs/specs/otel/metrics/) to a OTEL metrics collector or a metrics backend that support OTEL metrics fromat by using the [xk6-output-opentelemetry extension](https://github.com/grafana/xk6-output-opentelemetry/).

Reviewer: ```suggestion
Grafana k6 can push test run metrics in the [OpenTelemetry (OTEL) metrics format](https://opentelemetry.io/docs/specs/otel/metrics/) to an OTEL metrics collector or a metrics backend that supports the OTEL metrics format by using the [xk6-output-opentelemetry extension](https://github.com/grafana/xk6-output-opentelemetry/).
```

> This is work-in-progress extension, and we are looking for feedback and contributions. Feel free to open an issue directly in the [grafana/xk6-output-opentelemetry repository](https://github.com/grafana/xk6-output-opentelemetry/issues).

Reviewer 2: ```suggestion
This is an experimental extension, and we are looking for feedback and contributions. Feel free to open an issue directly in the [grafana/xk6-output-opentelemetry repository](https://github.com/grafana/xk6-output-opentelemetry/issues).
```

Author: I'd avoid using an experimental extension term until it becomes an experimental extension in k6 (hopefully v0.53) to avoid potential misunderstandings. Do you think this is a valid concern?

Reviewer 2: ```suggestion
This is an extension, and we are looking for feedback and contributions and aiming to graduate it as an experimental output within k6 core in one of the future releases. Feel free to open an issue directly in the [grafana/xk6-output-opentelemetry repository](https://github.com/grafana/xk6-output-opentelemetry/issues).
```
The part that I don't like is the fact we mark it as a work in progress which isn't anymore the reality, AFAIK. They can use and trust it.

I made a new suggestion, anyway, feel free to reject it if you don't like it.

> This is an extension, and we are looking for feedback and contributions and aiming to graduate it as an experimental output within k6 core in one of the future releases. Feel free to open an issue directly in the [grafana/xk6-output-opentelemetry repository](https://github.com/grafana/xk6-output-opentelemetry/issues).

Reviewer: ```suggestion
{{< admonition type="note" >}}

This is an extension, and we're looking for feedback and contributions and aiming to graduate it as an experimental output within Grafana k6 in a future release. Feel free to [open an issue](https://github.com/grafana/xk6-output-opentelemetry/issues) if you have any feedback or suggestions.

{{< /admonition >}}
```

Just a suggestion, I think we can switch this to an admonition and move it up to line 8 (between the H1 and the first paragraph) to make it more similar to how we use the caution admonition for experimental modules.

## Metrics mapping

> All [k6 metric types](https://grafana.com/docs/k6/<K6_VERSION>/using-k6/metrics) are converted into an equivalent [OpenTelemetry metric type](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#timeseries-model). The k6's tags converted to the OpenTelemetry attributes. Whenever it's possible, the units also passed with the metrics.

Reviewer: ```suggestion
k6 converts all [k6 metric types](https://grafana.com/docs/k6/<K6_VERSION>/using-k6/metrics) to an equivalent [OTEL metric type](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#timeseries-model), and all k6 tags to OTEL attributes. When possible, the units are also passed with the metrics.
```

I'm not 100% sure on the meaning of the last sentence, would you mind explaining it to me? 🤓

> | k6      | OpenTelemetry                                                                                                                                                                                                                                                                                                                                                                                      |

Reviewer: ```suggestion
| k6 Metric      | OpenTelemetry Metric                                                                                                                                                                                                                                                                                                                                                                                      |
```

| Counter | `Float64CounterOption`                                                                                                                                                                                                                                                                                                                                                                             |
| Gauge   | `Float64ObservableGauge`                                                                                                                                                                                                                                                                                                                                                                           |

> | Rate    | It splits into two `Int64Counter` counters with the names `metric_name.occurred` and `metric_name.total`. The `metric_name.occurred` counts only the number of non-zero occurencies, where the `metric_name.total` registers total number of positive and negative. We might reconsider it after the [xk6-output-opentelemetry#12](https://github.com/grafana/xk6-output-opentelemetry/issues/12). |

Reviewer: ```suggestion
| Rate    | Split into two `Int64Counter` counters named `metric_name.occurred` and `metric_name.total`. `metric_name.occurred` counts only the number of non-zero occurrences, and `metric_name.total` registers the total number of positive and negative occurrences. This might change in the future, refer to [xk6-output-opentelemetry#12](https://github.com/grafana/xk6-output-opentelemetry/issues/12) for more details. |
```

| Trend   | `Float64Histogram`                                                                                                                                                                                                                                                                                                                                                                                 |

## Build the k6 version

To build a k6 binary with the extension, first, make sure you have [Go](https://golang.org/doc/install) and [Git](https://git-scm.com/) installed on your machine.

Then, open your terminal and run the following commands:

```bash
# Install xk6
go install go.k6.io/xk6/cmd/xk6@latest

# Build the k6 binary
xk6 build --with github.com/grafana/xk6-output-opentelemetry

... [INFO] Build environment ready
... [INFO] Building k6
... [INFO] Build complete: ./k6
```

xk6 will create the new k6 binary in the local folder.

{{% admonition type="note" %}}

> To learn more about how to build custom k6 versions, check out [xk6](https://github.com/grafana/xk6).

Reviewer: ```suggestion
To learn more about how to build custom k6 versions, refer to [xk6](https://github.com/grafana/xk6).
```

{{% /admonition %}}

## Run the k6 test

Using the k6 binary you built in the previous step, you can use the `--out xk6-opentelemetry` option when running your tests to use this extension:

{{< code >}}

```bash
$ ./k6 run --out xk6-opentelemetry script.js
```

{{< /code >}}

## Configuration

The following options can be configured:

| Name                               | Value                                                                                                                                                           |
| ---------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `K6_OTEL_METRIC_PREFIX`            | Metric prefix. Default is empty.                                                                                                                                |
| `K6_OTEL_FLUSH_INTERVAL`           | How frequently to flush metrics from k6 metrics engine. Default is `1s`.                                                                                        |

> | `K6_OTEL_EXPORT_INTERVAL`          | Configures the intervening time between metrics exports. Default is `1s`.                                                                                       |

Reviewer 2: Related question, it sounds very low to be the default value, is it the default value in OTel world?

Reviewer 2: By intuition, I would say `10s` could be a compromise, but I didn't conduct any advanced research to provide a rational number here.
What I know is, that storages like Prometheus, not originally designed for push-based architecture, work better with a more relaxed interval.
Furthermore, as it could be used by free-tiers then a more relaxed interval could be also convenient for Grafana Cloud. Because for paid plans, we mostly charge per active series and not per sample. So, if we send 1 series with 1k pushes we are not charging more than 1 series with 1 sample, instead, it costs more on our side because we need to process more sent data.

Author: Let's define it as `10s` then, I'm fine with that for now :+1:

> | `K6_OTEL_HEADERS`                  | Configures headers in W3C Correlation-Context format without additional semi-colon delimited metadata (i.e. "k1=v1,k2=v2"). Passes the headers to the exporter. |

Reviewer: ```suggestion
| `K6_OTEL_HEADERS`                  | Configures headers in the W3C Correlation-Context format without additional semi-colon delimited metadata (i.e. "k1=v1,k2=v2"). Passes the headers to the exporter. |
```

| `K6_OTEL_GRPC_EXPORTER_INSECURE`   | Disables client transport security for the gRPC exporter.                                                                                                       |
| `K6_OTEL_GRPC_EXPORTER_ENDPOINT`   | Configures the gRPC exporter endpoint. Default is `localhost:4317`.                                                                                             |
| `K6_OTEL_HTTP_EXPORTER_INSECURE`   | Disables client transport security for the HTTP exporter.                                                                                                       |
| `K6_OTEL_HTTP_EXPORTER_ENDPOINT`   | Configures the HTTP exporter endpoint. Default is `localhost:4318`.                                                                                             |
| `K6_OTEL_HTTP_EXPORTER_URL_PATH`   | Configures the HTTP exporter path. Default is `/v1/metrics`.                                                                                                    |

> You could also use OpenTelemetry SDK configuration environment variables to configure the OpenTelemetry output, like the [gRPC exporter configuration](https://pkg.go.dev/go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc@v1.26.0). Still the `K6_OTEL_*` environment variables will take precedence over the OpenTelemetry SDK configuration environment variables.

Reviewer: ```suggestion
You can also use the OpenTelemetry SDK configuration environment variables to configure the OpenTelemetry output, like the [gRPC exporter configuration](https://pkg.go.dev/go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc@v1.26.0). The `K6_OTEL_*` environment variables take precedence over the OpenTelemetry SDK configuration environment variables.
```

| `K6_OTEL_METRIC_PREFIX`            | Metric prefix. Default is empty.                                                                                                                                |
| `K6_OTEL_FLUSH_INTERVAL`           | How frequently to flush metrics from k6 metrics engine. Default is `1s`.                                                                                        |
| `K6_OTEL_EXPORT_INTERVAL`          | Configures the intervening time between metrics exports. Default is `10s`.                                                                                      |
| `K6_OTEL_HEADERS`                  | Configures headers in the W3C Correlation-Context format without additional semi-colon delimited metadata (i.e. "k1=v1,k2=v2"). Passes the headers to the exporter. |
| `K6_OTEL_TLS_INSECURE_SKIP_VERIFY` | Disables server certificate verification.                                                                                                                       |

> | `K6_OTEL_TLS_CERTIFICATE`          | Configures the path to the certificate file for TLS credentials.                                                                                                |

Reviewer 2: ```suggestion
| `K6_OTEL_TLS_CERTIFICATE`          | Configures the path to the root CA certificate file for TLS credentials. If it is not provided but TLS is enabled then the host's root CAs set is used. |
```
